target/snitch_cluster/sw/snax/xdma/src/snax-xdma-lib.c

```diff
@@ -27,24 +26,28 @@ int32_t xdma_memcpy_nd(uint8_t* src, uint8_t* dst, uint32_t unit_size_src,
     csrw_ss(XDMA_DST_ADDR_PTR_LSB, (uint32_t)(uint64_t)dst);
     csrw_ss(XDMA_DST_ADDR_PTR_MSB, (uint32_t)((uint64_t)dst >> 32));
     // Rule check
-    // unit size only support 8 bytes or n * 64 bytes
-    XDMA_DEBUG_PRINT("unit size src: %d\n", unit_size_src);
-    XDMA_DEBUG_PRINT("unit size dst: %d\n", unit_size_dst);
-
-    if ((unit_size_src % 64 != 0) && (unit_size_src != 8)) {
-        XDMA_DEBUG_PRINT("unit size src error\n");
+    // The enabled spatial bound for input should be equal to the enabled
+    // spatial bound for output
+    if (bound_src[0] != bound_dst[0]) {
+        XDMA_DEBUG_PRINT(
+            "Enabled channels at src is not equal to enabled channels at "
+            "dst\n");
```

Comment:
Why is this the case? For example the following scenario: (when bitmasking is also available):
Copy 4 bytes from 8 channels at src, apply transpose, and write 8 bytes at 4 channels at dst. This should be possible I think?

Reply:
The system changes a little bit. Before this update, the spatial unrolling is done **automatically**, which means one dimension of stride and bound will be automatically unrolled and dispatched to different streamer channels. This requires one bound should be integer times of eight. 

As you need to disable some channels, it should be explicitly configured, instead of implicitly unrolled. Now, the innermost stride and bound is controlling this matter. If you set the bound smaller than 8(e.g. 2), then only first two streamer channels are turned on. 

Now the programming model is changed as follows: 
1. During the synthesis time, you still need to determine the total dimension of the xdma
2. The difference is one dimension is used by spatial thing, and other dimensions is used by spatial thing (which looks more like @xiaoling-yi 's version right)
3. Configuring the first dimension to numbers smaller than 8 to disable channels; configuring the first dimension to 8 (the maximal allowed numbers) to enable all channels. If, for instance, bitwidth is 1024b and narrow port is 32b, then the maximal number of this is 32. Chisel will automatically handle it. 
4. Configuring the first dimension to numbers greater than number of channels will not take effect. The xdma will still turn on all channels and **not** do automatic loop unrolling. 

I didn't design "mask" for streamer channel, as this design is not consistent with the definition of stride and bound we used before. The bit mask, which will introduce later today, will be a real mask😂

Reply:
Ok! I think I get it :). Only the part that src and dst must be equal channels I still don't get.
Can we not set separate bounds and strides for the innermostloop separately for source and destination streamer?

Reply:
Stride is of course possible! As you see there is no checking for stride in the code. 

For bound, yes it is actually also possible. However, the disabled channel will be padded with 0 only. For instance, if you set the innermost bound for the source at 0, and the innermost bound for the destination at 8, you will finally write 0 to destination (like memset).

Reply:
Ok, looks good! Thank you

Reply:
If you set the innermost bound for the source at 1, and the innermost bound for the destination at 8, you will finally write  the first channel with meaningful data, and other channels with 0. (Assuming you disable all the extensions)

If you play with extension like transpose, then zero goes to each bit of different channels. Then you need to use byte mask (coming out later) to mask these zeros.